### PR TITLE
out_skywalking: introduce Apache SkyWalking output plugin

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Rei Shimizu <rei@tetrate.io>

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

This PR introduces [Apache SkyWalking](https://skywalking.apache.org/) output plugin.
It consumes input buffer and format to SkyWalking logging style, then sends to log collection endpoint.  

It converts input logs as follows.

### Raw
```
This is the original log message
```

### Input (with tail in plugin)
```
{
   "log": "This is the original log message"
}
```

### Output (with Apache SkyWalking plugin)
```
[{
  "timestamp": 123456789,
  "service": "service", // This value is configured by user
  "serviceInstance": "instance", // This value is configured by user
  "body": {
    "json": {
      "json": "{\"log\": \"This is the original log message\"}"
    }
  } 
}]
```

We can see aggregated log messages on UI as follows.

Screenshot from 2021-06-14 16-04-36

cc @wu-sheng @kezhenxu94

### CODE REVIEW

**Comment by Shikugawa** [comment]
# Example Configuration
```
[INPUT]
    Name tail
    Path /var/log/syslog

[OUTPUT]
    name skywalking
    service_name sample_service
    instance_name sample_instance
```

# Debug Log
```
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/06/09 19:19:28] [ info] Configuration:
[2021/06/09 19:19:28] [ info]  flush time     | 5.000000 seconds
[2021/06/09 19:19:28] [ info]  grace          | 5 seconds
[2021/06/09 19:19:28] [ info]  daemon         | 0
[2021/06/09 19:19:28] [ info] ___________
[2021/06/09 19:19:28] [ info]  inputs:
[2021/06/09 19:19:28] [ info]      tail
[2021/06/09 19:19:28] [ info] ___________
[2021/06/09 19:19:28] [ info]  filters:
[2021/06/09 19:19:28] [ info] ___________
[2021/06/09 19:19:28] [ info]  outputs:
[2021/06/09 19:19:28] [ info]      skywalking.0
[2021/06/09 19:19:28] [ info] ___________
[2021/06/09 19:19:28] [ info]  collectors:
[2021/06/09 19:19:28] [ info] [engine] started (pid=3713048)
[2021/06/09 19:19:28] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2021/06/09 19:19:28] [debug] [storage] [cio stream] new stream registered: tail.0
[2021/06/09 19:19:28] [ info] [storage] version=1.1.1, initializing...
[2021/06/09 19:19:28] [ info] [storage] in-memory
[2021/06/09 19:19:28] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] flb_tail_fs_inotify_init() initializing inotify tail input
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] inotify watch fd=23
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] scanning path /tmp/hoge2
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] inode=15597590 with offset=1745 appended as /tmp/hoge2
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] scan_glob add(): /tmp/hoge2, inode 15597590
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] 1 new files found on path '/tmp/hoge2'
[2021/06/09 19:19:28] [debug] [skywalking:skywalking.0] created event channels: read=25 write=26
[2021/06/09 19:19:28] [debug] [output:skywalking:skywalking.0] configured sample_svc/sample_ins
[2021/06/09 19:19:28] [debug] [output:skywalking:skywalking.0] OAP address: 127.0.0.1:11800
[2021/06/09 19:19:28] [debug] [router] default match rule tail.0:skywalking.0
[2021/06/09 19:19:28] [ info] [sp] stream processor started
[2021/06/09 19:19:28] [debug] [input:tail:tail.0] inode=15597590 file=/tmp/hoge2 promote to TAIL_EVENT
[2021/06/09 19:19:28] [ info] [input:tail:tail.0] inotify_fs_add(): inode=15597590 watch_fd=1 name=/tmp/hoge2
[2021/06/09 19:19:35] [debug] [input:tail:tail.0] inode=15597590 events: IN_MODIFY 
[2021/06/09 19:19:38] [debug] [task] created task=0x7f208c011390 id=0 OK
[2021/06/09 19:19:38] [debug] [http_client] not using http_proxy for header
[2021/06/09 19:19:38] [debug] [http_client] header=POST /v3/logs HTTP/1.1
Host: 127.0.0.1:11800
Content-Length: 129
Content-Type: application/json
User-Agent: Fluent-Bit


[2021/06/09 19:19:38] [debug] [out coro] cb_destroy coro_id=0
[2021/06/09 19:19:38] [debug] [task] destroy task=0x7f208c011390 (task_id=0)
```

# Valgrind heap summary
```
[2021/06/09 19:17:59] [ info] [engine] service stopped
[2021/06/09 19:17:59] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=15597590 watch_fd=1
==3712892== 
==3712892== HEAP SUMMARY:
==3712892==     in use at exit: 0 bytes in 0 blocks
==3712892==   total heap usage: 788 allocs, 789 frees, 2,667,552 bytes allocated
==3712892== 
==3712892== All heap blocks were freed -- no leaks are possible
==3712892== 
==3712892== For lists of detected and suppressed errors, rerun with: -s
==3712892== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by Shikugawa** [comment]
@edsiper Thank you for this awesome project! I added a new output plugin for [Apache SkyWalking](https://skywalking.apache.org/). Could you take a look and any comments?

**Comment by hailin0** [comment]
Reference skywalking log data protocol. does the json content format need to be modified?

- [skywalking http-api](https://github.com/apache/skywalking/blob/master/docs/en/protocols/Log-Data-Protocol.md#http-api)
- [skywalking deserialization json log](https://github.com/apache/skywalking/blob/master/oap-server/server-receiver-plugin/skywalking-log-recevier-plugin/src/main/java/org/apache/skywalking/oap/server/recevier/log/provider/handler/rest/LogReportServiceRestHandler.java#L81)


image

31623253795_ pic_hd

File: plugins/out_skywalking/skywalking.c
Comment (hailin0) [inline plugins/out_skywalking/skywalking.c:79]: How to set oap address(host/port) through the configuration file?

Comment (Shikugawa) [reply plugins/out_skywalking/skywalking.c:79]: You can use port and host property as follows.

```
[OUTPUT]
    name skywalking
    service_name sample_service
    instance_name sample_instance
    host google.com
    port 80
```

**Comment by hailin0** [comment]
Are there docker images available for testing?

**Comment by Shikugawa** [comment]
> Are there docker images available for testing?

You can try docker by yourself. https://github.com/fluent/fluent-bit/tree/master/dockerfiles

**Comment by Shikugawa** [comment]
@edsiper friendly ping

File: plugins/out_skywalking/CMakeLists.txt
Comment (edsiper) [inline plugins/out_skywalking/CMakeLists.txt:2]: fix indentation

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:50]: auth_token is NULL

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:58]: leaking memory of ctx->auth_token

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:66]: "else" on next line please: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

Comment (edsiper) [reply plugins/out_skywalking/skywalking.c:66]: and all of the others...

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:99]: you are leaking all the memory allocated previously, it's better to create a destroy routine for your context

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:112]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:113]: bad indentation, it should be 4 spaces

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:127]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:177]: if out_body_str gets NULL the next line will crash it

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:219]: use flb_mp_count()

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:242]: fix indentation

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:294]: no printfs please

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:379]: please implement a config map for your configuration, reference: https://github.com/fluent/fluent-bit/blob/master/plugins/out_http/http.c#L349

**Review by edsiper — CHANGES_REQUESTED** [review]
I did a partial review, please fix all code styling and suggestions in the code, run it through valgrind to find possible memory leaks on exceptions, also please implement a config map for the configuration options.

**Comment by hailin0** [comment]
> > Are there docker images available for testing?
> 
> You can try docker by yourself. https://github.com/fluent/fluent-bit/tree/master/dockerfiles

Building docker image error

Can you help me?

macos big sur: 11.2.3

image

File: plugins/out_skywalking/skywalking.c
Comment (Shikugawa) [reply plugins/out_skywalking/skywalking.c:200]: added NULL check

File: plugins/out_skywalking/skywalking.c
Comment (Shikugawa) [reply plugins/out_skywalking/skywalking.c:391]: implemented with config map.

**Comment by Shikugawa** [comment]
@edsiper Thank you for reviewing! Could you proceed?

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:58]: this might lead to a crash, ctx is invalid

File: plugins/out_skywalking/skywalking.c
Comment (edsiper) [inline plugins/out_skywalking/skywalking.c:171]: flb_errno() before return.

**Review by edsiper — CHANGES_REQUESTED** [review]

**Comment by edsiper** [comment]
thanks minor changes requested. everybody else can confirm the connector works as expected ?

pls also submit a documentation page for fluent/fluent-bit-docs

File: plugins/out_skywalking/skywalking.c
Comment (Shikugawa) [reply plugins/out_skywalking/skywalking.c:171]: fixed

**Comment by Shikugawa** [comment]
> thanks minor changes requested. everybody else can confirm the connector works as expected ?
> 
> pls also submit a documentation page for fluent/fluent-bit-docs

@edsiper Thank you for reviewing. I added docs for this plugin. And I can follow up this plugin if users don't satisfy the current feature.

**Comment by Shikugawa** [comment]
nostale

**Comment by Shikugawa** [comment]
@edsiper Could you take a look please?

**Comment by nokute78** [comment]
Could you modify commit messages (except merge commit)
e.g. `out_skywalking: something`
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#commit-changes

- follow up
- review
- fix

**Comment by Shikugawa** [comment]
nostale

**Comment by Shikugawa** [comment]
> Could you modify commit messages (except merge commit) e.g. `out_skywalking: something` https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#commit-changes
> 
> * follow up
> * review
> * fix

@nokute78 @edsiper Sorry to keep waiting. I cleaned dirty commit messages.

**Comment by nokute78** [comment]
@Shikugawa Thank you for updating.
However your commit contains a file which is not related this patch. `lib/cmetrics/include/cmetrics/cmt_version.h`
Could you update commit again to remove the file ?

https://github.com/fluent/fluent-bit/pull/3592/files#diff-5e28c9b27ef6f9272c97835b5386e22e058b278aa2f95e3201b86fd224dd6b71

**Comment by Shikugawa** [comment]
@nokute78 Sorry. Fixed that

**Comment by nokute78** [comment]
@edsiper Could you check your request changes ?
I think @Shikugawa resolved all your request.

**Comment by nokute78** [comment]
@agup006 I think it is better to merge for v1.9.

By the way, if we merge master and not to merge 1.8 branch, it means that merging for v1.9.
I think current master branch is for v1.9.

**Comment by wu-sheng** [comment]
@agup006 @nokute78 Do you have more clear idea how this should continue?

**Comment by agup006** [comment]
We should be good to go here, need a final +1 from @edsiper or @leonardo-albertovich to merge. Then we can ship alongside 1.9.

Shouldwe do a parallel blog post alongside @wu-sheng ?

**Comment by wu-sheng** [comment]
Yes, I plan to work with @Shikugawa to post a blog. We will write one to post on skywalking.apache.org, is there a place of fluent to post about this?

**Comment by agup006** [comment]
Yes we can do fluentbit.io

**Comment by wu-sheng** [comment]
@Shikugawa Let's have some discussion about this.

**Comment by wu-sheng** [comment]
> Awesome! Yes we can do fluentbit.io

We are planning to write this, `Blog [SkyWalking x Fluent] Collect and Analysis Envoy Access Log through Fluent-bit and Apache SkyWalking`

**Comment by wu-sheng** [comment]
Anyone could merge this?

**Comment by leonardo-albertovich** [comment]
I probably won't have the time required to properly review this PR this week. At first glance I found some slight coding style issues and I have yet to read the configuration handling in depth (I thought I saw something there). If no one does it by then I'll take a look at it early next week.

**Comment by nokute78** [comment]
@wu-sheng @leonardo-albertovich 
I think this PR is blocked by @edsiper 's requesting to change.

@edsiper Could you check it ? I think it is already resolved.

**Comment by edsiper** [comment]
merged, thanks!

**Pull request merged**
